docs/proposals/kubectl-login.md

```diff
@@ -0,0 +1,233 @@
+
+The steps taken are:
+
+1. If no cluster configured, prompt user for cluster information.
+2. If no user is configured, discover the authentication strategies supported by the API Server.
+3. Prompt the user for some information based on the authentication strategy they choose.
+4. Attempt to login as a user, including authentication challenges such as OAuth2 flows, and display user info.
+
+Importantly, each of these steps only trigger if kubeconfig isn't configured
+(or is improperly configured). Users with fully configured kubeconfigs will only
+see the user they're logged in as, useful for opaque credentials such as X509
+certs or bearer tokens.
+
+The command differs from `kubectl config set-credentials` by:
+
+* Communicating with the API Server to determine if the user is supplying valid auth events.
+* Validating input and being opinionated about the input it asks for.
+* Triggering authentication challenges for example:
+  * Basic auth: Actually try to communicate with the API server.
+  * OpenID Connect: Create an OAuth2 redirect.
+
+
+## Server endpoints
+
+On the API Server side this command requires two non-resource paths.  First a
+path to discover the authentication strategies supported by the API Server.
+
+> GET /auth/authenticators
+
+```
+{
+  "kind": "AuthentactionModes",
+  "authenticators": [
+    {
+      "mode": "tokenfile"
+    },
+    {
+      "mode": "oidc",
+      "config": {
+        "issuer": "https://accounts.google.com",
+        "trusted-client-id": "foobar"
+      }
+    },
+    {
+      "mode": "keystone",
+      "config": {
+        "url": "https://keystone.example.com"
+      }
+    },
+  ]
+}
+```
+
+And another path to get the user a request is logged in as.
+
+> GET /auth/whoami
+
+```
+{
+  "kind": "UserInfo",
+  "name": "janedoe",
+  "groups": [
+    "admins",
+    "kube-system"
+  ]
+}
+```
+
+To support these endpoints, add two top level types to `pkg/api/unversioned`.
+
+```go
+// AuthenticationModes lists the authentication strategies available, to allow
+// clients to discover which strategies are valid.
+type AuthenticationModes struct {
+	TypeMeta `json:",inline"`
+
+	// The list of authentication strategies enabled on this API server.
+	Authenticators []Authenticator `json:"authenticators"`
+}
+
+
+// Authenticator is a strategy for authenticating with the Kubernetes API.
+type Authenticator struct {
+        // The type of the authenticator such as "tokenfile" or "webhook".
+	Mode string `json:"mode"`
+
+	// Config contains public fields which indicate the configuration of
+        // the plugin. This is helpful to validate fields.
+	Config map[string]string `json:"config"`
+}
+
+// UserInfo represents the user information associated with a request, to allow
+// clients to determine the identity their credentials resolve to.
+type UserInfo struct {
+	TypeMeta `json:",inline"`
+
+        // String that uniquely identifies the user.
+	Name string `json:"name"`
```

Comment:
This type actually already exists in `authentication.k8s.io/v1beta1`[0]. Going to reuse that.

[0] https://github.com/kubernetes/kubernetes/blob/8780b45a0486ea53e2bae58bbb6bd7c8a1cbc42c/pkg/apis/authentication.k8s.io/v1beta1/types.go#L52